### PR TITLE
perf: bound DWARF DIE-cache memory on large binaries

### PR DESCRIPTION
**Pull request by napetrov**
## Summary

`dump` on a large template-heavy C++ library (e.g. oneAPI DAAL, ~108 MB `.debug_info`) was measured peaking above 12-16 GB of RSS and taking minutes, close to/over a 16 GB CI runner's headroom.

## Motivation

pyelftools permanently caches every DIE it parses inside a `CompileUnit`, for the lifetime of the `DWARFInfo` object — there's no eviction. abicheck's `DwarfSession` (`dwarf_unified.py`) deliberately shares one `DWARFInfo` across up to three full-tree DWARF walks (basic metadata + advanced metadata, snapshot build, layout backfill) so later passes hit that cache instead of re-parsing (an intentional prior optimization). For a small library this is a clear win. For a binary with millions of DIEs, the per-DIE Python object overhead (each DIE holds an attributes dict of `AttributeValue` records plus bookkeeping) inflates that permanently-cached tree to roughly 50-100x the raw `.debug_info`/`.debug_loc` section bytes, which matches the observed multi-GB RSS from ~150 MB of raw DWARF sections.

## Changes

- `abicheck/dwarf_utils.py`: new `free_cu_die_cache()` (drops a `CompileUnit`'s cached DIE objects — safe, since pyelftools transparently re-parses from the stream on the next access) and `dwarf_low_memory_mode()` (decides from `.debug_info` section size vs. a threshold, default 32 MiB, overridable via `ABICHECK_DWARF_LOW_MEMORY_MB`, whether a walk should free eagerly).
- Wired into the two production DWARF walkers: `dwarf_unified.parse_dwarf_from_session` (basic + advanced metadata pass) and `dwarf_snapshot._DwarfSnapshotBuilder._walk_dwarf` (the snapshot-build walk, which also backs the layout-backfill path via `dumper_layout_backfill.py`). Each frees a CU's DIE cache immediately after finishing with it when in low-memory mode, bounding peak memory to roughly one CU's tree instead of the whole binary's, trading the cross-pass cache-reuse CPU savings for bounded peak memory on binaries large enough for that trade to matter.
- Small binaries (the common case) are unaffected — the default threshold keeps the existing fast cache-reuse behavior unchanged.
- `changelog.d/`: performance fragment.

## Tests

- `tests/test_dwarf_low_memory.py`: unit tests for the cache-freeing/threshold mechanics (including env-var override and non-numeric/mocked-size edge cases), plus Linux/gcc integration tests proving low-memory mode is byte-for-byte output-neutral against `parse_dwarf`/`build_snapshot_from_dwarf` in both the session-reuse and direct-open code paths.
- Full fast suite: `23210 passed, 13 skipped, 4 xfailed` (0 failures).
- `mypy abicheck/`: clean. `ruff check`: clean on touched files.

## Checklist

- [x] Tests added / updated for new behaviour
- [x] Changelog fragment added (`scriv create`)
- [ ] Docs updated if user-visible behaviour changed — no user-facing CLI/API surface changed; the new `ABICHECK_DWARF_LOW_MEMORY_MB` env var is an internal tuning knob mirroring existing ones (`ABICHECK_TU_JOBS`, etc.)
- [x] No new `mypy` or `ruff` errors introduced

---
_Generated by [Claude Code](https://claude.ai/code/session_01EPw9kJTsSV2TCNTjkdxXWY)_

### CODE REVIEW

File: abicheck/dwarf_utils.py [hunk 1]
```diff
@@ -288,3 +292,93 @@ def resolve_type_die(die: Any, CU: Any) -> Any | None:
...
+    large binary.
+
+    No-ops if *CU* doesn't carry the private pyelftools cache attributes (a
+    version mismatch, or a CU that hasn't cached anything yet) rather than
+    raising -- freeing is always best-effort.
+    """
+    if hasattr(CU, "_dielist"):
+        CU._dielist = []
+    if hasattr(CU, "_diemap"):
+        CU._diemap = []
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/dwarf_utils.py]: Fixed in 916e403: `free_cu_die_cache` now calls `.clear()` in place on both containers instead of reassigning `[]`, so it degrades safely regardless of the container type (checked pyelftools 0.29 and 0.33 directly — both use parallel lists, but `.clear()` removes the assumption either way).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: abicheck/dwarf_utils.py [hunk 2]
```diff
@@ -288,3 +292,93 @@ def resolve_type_die(die: Any, CU: Any) -> Any | None:
...
+            threshold_mb = DEFAULT_DWARF_LOW_MEMORY_THRESHOLD_MB
+    debug_info_sec = getattr(dwarf_info, "debug_info_sec", None)
+    try:
+        size = int(getattr(debug_info_sec, "size", 0) or 0)
+    except (TypeError, ValueError):
+        # A test double (e.g. MagicMock) or otherwise non-numeric ``.size``
+        # carries no real size signal — treat as unknown/small rather than
+        # raise, matching every other "never raise" DWARF helper here.
+        size = 0
+    return (size / (1024 * 1024)) > threshold_mb
```
[the remark this answers is not in the document]
Comment (napetrov) [reply abicheck/dwarf_utils.py:367]: Good catch — fixed in 916e403: `dwarf_low_memory_mode` now short-circuits to `False` for any negative threshold before the size comparison, matching the documented disable semantics. Added a regression test (`test_negative_env_override_disables_low_memory_mode`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_dwarf_low_memory.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+
+# ---------------------------------------------------------------------------
+# End-to-end: low-memory mode must not change any parsed output
+# ---------------------------------------------------------------------------
+
+@pytest.mark.skipif(
+    sys.platform != "linux",
+    reason="ELF DWARF tests require Linux (macOS/Windows compilers produce Mach-O/PE)",
+)
+class TestLowMemoryModeIsOutputNeutral:
```
[the remark this answers is not in the document]
Comment (napetrov) [reply tests/test_dwarf_low_memory.py:198]: Not applying this one — it would diverge from the existing convention in this exact test family. `tests/test_dwarf_unified.py` (which these tests were deliberately modeled on) has several compiler-backed classes (`TestUnifiedEqualsSepaRate`, `TestShims`, `TestSingleOpen`, `TestDwarfSession`) that also compile real `.so`s with g++ via the same `_compile_so`/`_require_tool` helpers, gated only by the platform `skipif` — none carry `pytest.mark.integration`. They already run in the default fast lane today (confirmed: `pytest tests/ -k dwarf -m "not integration and not ..."` collects and passes them, gcc/g++ being ordinary dev/CI-image tooling unlike castxml/abidiff/abi-compliance-checker, which the `integration`/`libabigail`/`abicc` markers exist to gate). Marking only the new tests `integration` would make identical-shaped tests in the same suite behave inconsistently for no functional reason. Happy to revisit if there's a reason to change the whole family's convention, but that's a separate, broader decision than this PR.

---
_Generated by [Claude Code](https://claude.ai/code)_
